Mathlib/RingTheory/MvPolynomial/NewtonIdentities.lean

```diff
@@ -0,0 +1,320 @@
+/-
+Copyright (c) 2023 Michael Lee. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Lee
+-/
+import Mathlib.Algebra.Algebra.Subalgebra.Basic
+import Mathlib.Data.Finset.Card
+import Mathlib.Data.Fintype.Basic
+import Mathlib.Data.MvPolynomial.CommRing
+import Mathlib.Data.MvPolynomial.Rename
+import Mathlib.Data.Nat.Parity
+import Mathlib.RingTheory.MvPolynomial.Symmetric
+
+/-!
+# Newton's Identities
+
+This file defines `MvPolynomial` power sums as a means of implementing Newton's identities.
+
+## Main declarations
+
+* `MvPolynomial.psum`
+
+* `MvPolynomial.NewtonIdentity`
+
+## Notation
+
++ `psum σ R n` is the degree-`n` power sum in `MvPolynomial σ R`.
+
+As in other polynomial files, we typically use the notation:
+
++ `σ τ : Type _` (indexing the variables)
+
++ `R S : Type _` `[CommSemiring R]` `[CommSemiring S]` (the coefficients)
+
++ `r : R` elements of the coefficient ring
+
++ `i : σ`, with corresponding monomial `X i`, often denoted `X_i` by mathematicians
+
++ `φ ψ : MvPolynomial σ R`
+
+## References
+
+See [zeilberger1984] for the combinatorial proof of Newton's identities.
+-/
+
+open Equiv (Perm)
+
+open BigOperators MvPolynomial
+
+noncomputable section
+
+section PowerSum
+
+namespace Multiset
+
+variable {R : Type _} [CommSemiring R]
+
+/-- The degree-`n` power sum evaluated at the elements of `s` -/
+def psum (s : Multiset R) (n : ℕ) : R := (s.map (fun (x : R) ↦ x ^ n)).sum
+
+end Multiset
+
+open Finset
+
+variable (σ R) [CommSemiring R] [Fintype σ] [Fintype τ]
+
+/-- The degree-`n` power sum -/
+def psum (n : ℕ) : MvPolynomial σ R := ∑ i in univ, (X i) ^ n
+
+theorem rename_psum (n : ℕ) (e : σ ≃ τ) : rename e (psum σ R n) = psum τ R n := by
+  calc
+    rename e (psum σ R n) = ∑ i in univ, (X (e i)) ^ n := by
+      simp_rw [psum, map_sum, map_pow, rename_X]
+    _ = ∑ i in univ.map e.toEmbedding, (X i) ^ n := by simp [-map_univ_equiv]
+    _ = ∑ i in univ, (X i) ^ n := by rw [map_univ_equiv]
+
+theorem psum_isSymmetric (n : ℕ) : IsSymmetric (psum σ R n) := by
+  intro
+  rw [rename_psum]
+
+end PowerSum
+
+section Newton
+
+open Classical Finset Nat
+
+variable (σ : Type) [Fintype σ] [DecidableEq σ] [Fintype τ] (R : Type) [CommRing R]
+  [NoZeroDivisors (MvPolynomial σ R)] [CharZero (MvPolynomial σ R)]
+/-
+  TODO: show that MvPolynomial σ R is an integral domain if R is an integral domain
```

Comment:
I think we have this result? Maybe I'm misremembering.

Reply:
I couldn't get `NoZeroDivisors (MvPolynomial σ R)` to synthesize after adding `[NoZeroDivisors R]`. I might have done this incorrectly, though.

Reply:
It's possible we have something only for `IsDomain`.

Reply:
https://leanprover-community.github.io/mathlib_docs/ring_theory/polynomial/basic.html#mv_polynomial.no_zero_divisors seems to exist (excuse the Lean3 docs)

Reply:
we also have the `char_p` result, but not for `char_zero`: https://leanprover-community.github.io/mathlib_docs/ring_theory/mv_polynomial/basic.html#mv_polynomial.char_p

Reply:
Very nice, thanks for finding! I'm not sure why I'm getting `synthInstance` errors. Will investigate further.

Reply:
Okay, just needed to import Mathlib.RingTheory.Polynomial.Basic for the first one. Writing up an `instance` for `CharZero` for the other.